File: docs/exec-plans/evidence/001/studies/gate-0/validate.rb
require "digest"
require "json"

dir = __dir__
canonical = lambda do |value|
  case value
  when Hash then "{" + value.keys.sort.map { |key| JSON.generate(key) + ":" + canonical.call(value[key]) }.join(",") + "}"
  when Array then "[" + value.map { |item| canonical.call(item) }.join(",") + "]"
  else JSON.generate(value)
  end
end
hash_object = ->(value) { Digest::SHA256.hexdigest(canonical.call(value)) }
read_canonical = lambda do |name|
  raw = File.binread(File.join(dir, name)); value = JSON.parse(raw)
  abort "gate-0 Ruby validator: noncanonical #{name}" unless raw == canonical.call(value) + "\n"
  value
end
snapshot = read_canonical.call("snapshot.jcs.json")
oracle = read_canonical.call("observable-gate.jcs.json")
abort "gate-0 Ruby validator: identity" unless snapshot["schemaVersion"] == "eonfolk-gate0-snapshot-v1" && snapshot["gateId"] == "gate-0"
answer = snapshot["answerKey"]
expected_answer = { "mara"=>"citizen:mara", "activities"=>"activity:carry-water,activity:exchange-rations,activity:gather-wood", "interaction"=>"interaction:iven,toma|exchange-settled", "autonomy"=>"cannot-command|standing-plan" }
abort "gate-0 Ruby validator: answer key" unless answer == expected_answer
rows = snapshot.dig("treatmentSnapshot", "counterbalance")
abort "gate-0 Ruby validator: rows" unless rows.length == 6 && rows.all? { |row| row.sort == %w[DIR ECH FAC FAM H TRI] }
pairs = rows.flat_map { |row| row.each_cons(2).to_a }
abort "gate-0 Ruby validator: predecessor pairs" unless pairs.uniq.length == 30
actions = %w[verify-private accuse-now abstain]
treatment_snapshot = snapshot["treatmentSnapshot"]
treatments = treatment_snapshot["treatments"]
vectors = treatment_snapshot["vectors"]
abort "gate-0 Ruby validator: vector cardinality" unless vectors.length == 18 && vectors.map { |v| [v["treatmentId"], v["adviceInput"]] }.uniq.length == 18
vectors.each do |vector|
  definition = treatments.fetch(vector["treatmentId"])
  scores = vector["chooserScores"]
  abort "gate-0 Ruby validator: voter list" unless scores.map { |score| score["chooser"] } == definition["voters"]
  scores.each do |score|
    maximum = score["scores"].max
    expected_preference = actions[score["scores"].index(maximum)]
    abort "gate-0 Ruby validator: chooser preference" unless score["preferredAction"] == expected_preference
  end
  totals = actions.to_h { |action| [action, scores.count { |score| score["preferredAction"] == action }] }
  abort "gate-0 Ruby validator: vote totals" unless totals == vector["voteTotals"]
  chosen = if definition["rule"] == "direct"
    vector["adviceInput"]
  elsif definition["rule"] == "individual"
    scores.first.fetch("preferredAction")
  else
    maximum = totals.values.max
    tied = actions.select { |action| totals[action] == maximum }
    if tied.length == 1 || definition["rule"] == "trio"
      tied.first
    else
      mara = scores.find { |score| score["chooser"] == "Mara" }
      tied.max_by { |action| [mara.fetch("scores")[actions.index(action)], -actions.index(action)] }
    end
  end
  abort "gate-0 Ruby validator: resolved choice" unless chosen == vector["chosenAction"]
  disposition = if definition["rule"] == "direct" then "commanded"
    elsif vector["adviceInput"] == chosen then "accepted"
    elsif chosen == "abstain" then "delayed"
    elsif vector["adviceInput"] == "abstain" then "rejected"
    else "reinterpreted" end
  abort "gate-0 Ruby validator: disposition" unless disposition == vector["disposition"]
  abort "gate-0 Ruby validator: terminal hash" unless Digest::SHA256.hexdigest(canonical.call(vector["terminalState"])) == vector["terminalStateHash"]
  abort "gate-0 Ruby validator: consequence key" unless vector["renderedConsequenceKey"] == "gate-0.consequence.#{chosen}"
end
inputs = snapshot["oracleInputs"]
expected = {
  "acceptedTreatmentHash"=>hash_object.call(inputs["acceptedTreatment"]), "acceptedTreatmentId"=>"H",
  "anchorsHash"=>hash_object.call(inputs["anchors"]), "chronicleHash"=>hash_object.call(snapshot.dig("visualFixture", "chronicleBeat")),
  "comparisonContractHash"=>hash_object.call(inputs["comparisonContract"]), "eventIntervalHash"=>nil,
  "fixtureHash"=>hash_object.call(inputs["fixture"]), "gateId"=>"gate-0", "logicalTimelineHash"=>hash_object.call(inputs["logicalTimeline"]),
  "optionOrdersHash"=>hash_object.call(inputs["optionOrders"]), "optionSetsHash"=>hash_object.call(inputs["optionSets"]),
  "questionsHash"=>hash_object.call(inputs["questions"]), "readyPredicateHash"=>hash_object.call(inputs["readyPredicate"]),
  "receiptHash"=>nil, "rendererMode"=>"pixi-semantic", "responseSurfaceHash"=>hash_object.call(inputs["responseSurface"]),
  "routeId"=>"gate-0-combined-study", "routeParams"=>{"observer"=>{"capture"=>"1","fixtureId"=>"gate0-visual-v1","studyId"=>"V01"},"product"=>{"studyId"=>"P01"}},
  "schemaVersion"=>"eonfolk-observable-gate-v1", "scriptHash"=>hash_object.call(inputs["script"]),
  "semanticDomHash"=>hash_object.call(inputs["semanticDom"]), "stateHash"=>hash_object.call(snapshot["region"]), "storyCardHash"=>nil,
  "timersHash"=>hash_object.call(inputs["timers"]), "viewportPngHashes"=>{}
}
%w[desktop-1728x1117 laptop-1366x768 mobile-390x844].each { |id| expected["viewportPngHashes"][id] = Digest::SHA256.file(File.join(dir, "viewports", id + ".png")).hexdigest }
abort "gate-0 Ruby validator: observable oracle" unless expected == oracle
labor = snapshot["laborAdmission"]
abort "gate-0 Ruby validator: labor" unless labor["recordedNonOperatorSeconds"] + labor["openOperatorReservationSeconds"] + labor["protectedMappedHighSeconds"] <= labor["mappedBudgetSeconds"]
puts "gate-0 Ruby artifact valid: oracle #{Digest::SHA256.file(File.join(dir, "observable-gate.jcs.json")).hexdigest}"

File: docs/exec-plans/evidence/001/studies/gate-0/validate.mjs
import { createHash } from "node:crypto";
import { readFile } from "node:fs/promises";
import { TREATMENT_SNAPSHOT } from "../../../../../../tests/prototypes/gate-0/product/contract.ts";
import {
	GATE0_ANCHORS,
	GATE0_ASSIGNMENTS,
	GATE0_OPERATIONAL_SEED,
	GATE0_OPTION_SETS,
	GATE0_QUESTIONS,
	GATE0_SCRIPT,
	GATE0_TIMERS_MS,
	OPERATOR_FOCUSED_MINUTES_CAP,
} from "../../../../../../tests/prototypes/gate-0/study/contract.ts";
import { gate0OptionOrders } from "../../../../../../tests/prototypes/gate-0/study/prng.ts";
import {
	buildGate0StudyDefinition,
	validateGate0Evidence,
	validateGate0StudyDefinition,
	validateObserverAnswerKey,
} from "../../../../../../tests/prototypes/gate-0/study/validate.ts";
import {
	ANSWER_KEY,
	CHRONICLE_BEAT,
	FIXTURE,
} from "../../../../../../tests/prototypes/gate-0/visual/fixture.mjs";

const sha256 = (bytes) => createHash("sha256").update(bytes).digest("hex");
const canonical = (value) =>
	value === null || typeof value !== "object"
		? JSON.stringify(value)
		: Array.isArray(value)
			? `[${value.map(canonical).join(",")}]`
			: `{${Object.keys(value)
					.sort()
					.map((key) => `${JSON.stringify(key)}:${canonical(value[key])}`)
					.join(",")}}`;
const hashObject = (value) => sha256(Buffer.from(canonical(value)));
const equal = (a, b) => canonical(a) === canonical(b);
const assert = (condition, message) => {
	if (!condition) throw new Error(`gate-0 artifact violation: ${message}`);
};
const dir = new URL("./", import.meta.url);
const readCanonical = async (name) => {
	const raw = await readFile(new URL(name, dir));
	const value = JSON.parse(raw);
	assert(
		raw.equals(Buffer.from(`${canonical(value)}\n`)),
		`${name} is not RFC8785+LF`,
	);
	return { raw, value };
};

export async function validateArtifact() {
	const { value: snapshot } = await readCanonical("snapshot.jcs.json");
	const { value: oracle } = await readCanonical("observable-gate.jcs.json");
	assert(
		snapshot.schemaVersion === "eonfolk-gate0-snapshot-v1" &&
			snapshot.gateId === "gate-0",
		"snapshot identity",
	);
	assert(
		validateGate0StudyDefinition(snapshot.studyDefinition).ok &&
			equal(snapshot.studyDefinition, buildGate0StudyDefinition()),
		"study definition",
	);
	assert(
		validateObserverAnswerKey(snapshot.answerKey).ok &&
			equal(snapshot.answerKey, ANSWER_KEY),
		"answer key",
	);
	assert(
		equal(snapshot.treatmentSnapshot, TREATMENT_SNAPSHOT),
		"treatment snapshot",
	);
	assert(equal(snapshot.visualFixture, FIXTURE), "visual fixture");
	assert(
		snapshot.laborAdmission.recordedNonOperatorSeconds +
			snapshot.laborAdmission.openOperatorReservationSeconds +
			snapshot.laborAdmission.protectedMappedHighSeconds <=
			snapshot.laborAdmission.mappedBudgetSeconds,
		"labor admission",
	);
	const inputs = snapshot.oracleInputs;
	assert(
		equal(inputs.optionOrders, gate0OptionOrders(GATE0_OPERATIONAL_SEED)),
		"option orders",
	);
	assert(
		equal(inputs.optionSets, GATE0_OPTION_SETS) &&
			equal(inputs.script, GATE0_SCRIPT) &&
			equal(inputs.questions, GATE0_QUESTIONS) &&
			equal(inputs.anchors, GATE0_ANCHORS) &&
			equal(inputs.timers, GATE0_TIMERS_MS),
		"instrument inputs",
	);
	const viewportPngHashes = {};
	for (const id of ["desktop-1728x1117", "laptop-1366x768", "mobile-390x844"])
		viewportPngHashes[id] = sha256(
			await readFile(new URL(`viewports/${id}.png`, dir)),
		);
	const expected = {
		acceptedTreatmentHash: hashObject(inputs.acceptedTreatment),
		acceptedTreatmentId: "H",
		anchorsHash: hashObject(inputs.anchors),
		chronicleHash: hashObject(CHRONICLE_BEAT),
		comparisonContractHash: hashObject(inputs.comparisonContract),
		eventIntervalHash: null,
		fixtureHash: hashObject(inputs.fixture),
		gateId: "gate-0",
		logicalTimelineHash: hashObject(inputs.logicalTimeline),
		optionOrdersHash: hashObject(inputs.optionOrders),
		optionSetsHash: hashObject(inputs.optionSets),
		questionsHash: hashObject(inputs.questions),
		readyPredicateHash: hashObject(inputs.readyPredicate),
		receiptHash: null,
		rendererMode: "pixi-semantic",
		responseSurfaceHash: hashObject(inputs.responseSurface),
		routeId: "gate-0-combined-study",
		routeParams: {
			observer: { capture: "1", fixtureId: "gate0-visual-v1", studyId: "V01" },
			product: { studyId: "P01" },
		},
		schemaVersion: "eonfolk-observable-gate-v1",
		scriptHash: hashObject(inputs.script),
		semanticDomHash: hashObject(inputs.semanticDom),
		stateHash: hashObject(FIXTURE.region),
		storyCardHash: null,
		timersHash: hashObject(inputs.timers),
		viewportPngHashes,
	};
	assert(equal(oracle, expected), "observable oracle");
	return {
		snapshotHash: sha256(await readFile(new URL("snapshot.jcs.json", dir))),
		observableGateHash: sha256(
			await readFile(new URL("observable-gate.jcs.json", dir)),
		),
		assignments: GATE0_ASSIGNMENTS.length,
		operatorFocusedMinutesCap: OPERATOR_FOCUSED_MINUTES_CAP,
	};
}

export async function validateEvidence(path, studyCommit, manifestHash) {
	const result = validateGate0Evidence(await readFile(path), {
		studyCommit,
		manifestHash,
	});
	assert(result.ok, result.errors.join("; "));
	return result;
}

if (process.argv[1] === new URL(import.meta.url).pathname)
	process.stdout.write(`${JSON.stringify(await validateArtifact())}\n`);
